### PR TITLE
Add masked select api

### PR DESCRIPTION
**Pull request by shippingwang**
It selects elements of the input tensor according to the mask tensor.
Note: The shapes of the mask tensor don't have to match the shapes of input tensor, but they must be broadcastable, and the result is a new 1-D tensor.

NOTE: As the broadcast function is not elegantly implemented in PaddlePaddle, we assume the broadcastable concept in here is consistent with [**api:expand_as**](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api_cn/layers_cn/expand_as_cn.html#expand-as)

image

### CODE REVIEW

**Review by heavengate — DISMISSED** [review]

**Review by wanghaoshuang — DISMISSED** [review]

**Comment by shippingwang** [comment]
https://github.com/PaddlePaddle/FluidDoc/pull/1610 中文预览PR

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -13458,3 +13459,63 @@ def get_attr_shape(list_shape):
...
+def masked_select(input, mask):
+    """
+    This OP selects elements of the input tensor according to the mask tensor.
+    The shapes of the mask tensor don't have to match shapes of input tensor, but they must be broadcastable, and the result is a new 1-D tensor.
+
+    NOTE: The meaning of broadcastable is consistent with expand_as.
+
+    Parameters:
+
+        input(Variable): The input tensor, the data type should be float32.
```
Comment (lanxianghit) [inline python/paddle/fluid/layers/nn.py]: 为什么只支持float32

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -13458,3 +13459,63 @@ def get_attr_shape(list_shape):
...
+        .. code-block:: python
+
+            import paddle.fluid as fluid
+            import numpy as np
+            mask_shape = [4,1]
+            shape = [4,4]
+            data = np.random.random(mask_shape).astype("float32")
+            input_data = np.random.randint(5,size=shape).astype("float32")
+            mask_data = data > 0.5
+
```
Comment (lanxianghit) [inline python/paddle/fluid/layers/nn.py:13489]: 把data和mask_data也打印一下

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -13458,3 +13459,63 @@ def get_attr_shape(list_shape):
...
+            input = fluid.data(name="input",shape=[4,4],dtype="float32")
+            mask = fluid.data(name="mask",shape=[4,1],dtype="bool")
+            result = fluid.layers.masked_select(input=input, mask=mask)
+            place = fluid.CPUPlace()
+            exe = fluid.Executor(place)
+            start = fluid.default_startup_program()
+            main = fluid.default_main_program()
+            exe.run(start)
+            masked_select_result= exe.run(main, feed={'input':input_data, 'mask':mask_data}, fetch_list=[result])
+            # print(masked_select)
```
Comment (lanxianghit) [inline python/paddle/fluid/layers/nn.py]: print(masked_select)
变量名写错了

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -13458,3 +13459,63 @@ def get_attr_shape(list_shape):
...
+            # [0.38972723 0.36218056 0.7892614  0.50122297 0.14408113 0.85540855
+            #   0.30984417 0.7577004  0.26214206 0.32359877 0.6314582  0.2128865 ]
+
+    """
+    mask_cast = cast(x=mask, dtype="float32")
```
Comment (lanxianghit) [inline python/paddle/fluid/layers/nn.py]: 两次cast是因为expand_as不支持bool类型？如果是这样，我觉得应该提需求升级expand_as

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py:13489]: > 把data和mask_data也打印一下

mask_data打印了，data只是用来造mask的，没有给出。

File: python/paddle/fluid/layers/nn.py (hunk 3 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py]: > print(masked_select)
> 变量名写错了

嗯呢

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py]: > 两次cast是因为expand_as不支持bool类型？如果是这样，我觉得应该提需求升级expand_as

不是不支持bool型，是现在expand_as要求输入输出的类型要一样, 所以前后添加了cast

File: python/paddle/fluid/layers/nn.py (hunk 1 above)
Comment (shippingwang) [reply python/paddle/fluid/layers/nn.py]: > 为什么只支持float32

OK 我改一下，这里需要获取输入的dtype传给cast

**Review by lanxianghit — APPROVED** [review]
expand_as设计不合理的问题请提P0卡片需求

**Review by Boyan-Liu — APPROVED** [review]
LGTM

**Pull request merged**